### ISSUE TITLE
Max tile size guidelines?

### ISSUE THREAD

**Issue by savardge** [zpzim/SCAMP#14]
Hi Zach, 

thanks for making SCAMP available! I'm a seismologist at U of Calgary and SCAMP has been successful in helping me with challenging earthquake detection in my datasets, which is awesome!

I have a question regarding the max tile size parameter. I run SCAMP on my university computer cluster, using two nodes with two Tesla V100-PCIE-16GB GPUs each. In some cases, I've had the run terminate due to an "out-of-memory" exception, which made me think I had to change the max tile size parameter. Because I'm relatively new to GPU computing, I was wondering if you would have any tips or guidelines on how to set up this parameter for a given dataset size and given GPU memory available?

Thanks in advance!

**Comment by zpzim** [comment]
Hi there! It's great to see more seismologists taking advantage of SCAMP! I know firsthand SCAMP can do a lot for that domain and am excited to see new and interesting ways it can help.

As for your problem: it is unlikely SCAMP is using all 16GB of memory on each of your GPUs.

Can you give me a few more bits of information?

1. Input size and configuration, can you send me the entire set of command line arguments you are running with as well as the number of datapoints in your inputs.

2. How are you running on both nodes at the same time? Are you using the distributed configuration with the clients/server or the standard one, maybe with SLURM or something?

3.  How much system memory is available on those cluster nodes.

4. Do you have exclusive use of those machines? Or can other people run code on them at the same time?

5. Can you send me the actual text of the error message?

Thanks!
Zach

**Comment by savardge** [comment]
Of course, thanks for your time! Please see below for an example run:

> 1. Input size and configuration, can you send me the entire set of command line arguments you are running with as well as the number of datapoints in your inputs.

My input file is 4GB in size, with 167,646,039 entries. Here is my slurm batch job submission script and SCAMP command arguments, in this example, I use 2 nodes with 4 GPUs each:

```
#!/bin/bash

#SBATCH --partition=glados-gpu
#SBATCH --gres=gpu:4
#SBATCH --nodes=2
#SBATCH --time=7-00:00:00
#SBATCH --output=501_DP1_all_filt2-10.log

pattern="501_DP1_all_filt2-10"
winlen=50

win="--window=${winlen}"
inputa="--input_a_file_name=501_DP1_all_filt2-10.dat"
dblprec="--double-precision=true"
outmp="--output_a_file_name=${pattern}_win${winlen}.mp"
outind="--output_a_index_file_name=${pattern}_win${winlen}.mpind"
pearson="--output_pearson=true"


/home/genevieve.savard/SCAMP/build/SCAMP $win $inputa $dblprec $outmp $outind $pearson
```

> 2. How are you running on both nodes at the same time? Are you using the distributed configuration with the clients/server or the standard one, maybe with SLURM or something?

I'm using SLURM as per the above script.

> 3. How much system memory is available on those cluster nodes.

In the example above, I use 2 nodes each with: 
- 20 cores/node (two 2.4 GHz Intel Xeon CPU E5-2640 V4 processors) 
- 250 GB total memory (per node)
- Four GeForce GTX 1080 Ti (Pascal) GPUs each with 11 GB of memory (per node)

> 4. Do you have exclusive use of those machines? Or can other people run code on them at the same time?

Unfortunately not, these nodes are shared with other researchers.

> 5. Can you send me the actual text of the error message?

The error message in the slurm output is: `GPUasssert: out of memory /home/genevieve.savard/SCAMP/src/tile.cpp 151`
The job was terminated after 2 days and 6 hours of running.
Is it possible it ran out of memory at the moment of writing the output files? Does it use the CPU at that moment? I didn't specify explicitly CPU memory in previous runs where I had a bigger input, so I'm a bit puzzled why these runs wouldn't work. Maybe something to do with slurm accounting parameters too?

Let me know your thoughts!

**Comment by zpzim** [comment]
Hmm this is puzzling. I preallocate almost all memory beforehand. And you have 250GB of memory on each node so system memory should be fine. The failure is on a call to cudaMalloc()  which is even more strange to have that be far into the run. It also could be that the failure is somewhere else and it just surfaced there for some reason. 

I am thinking this could be related to something SLURM is doing (like somehow a bunch memory is being by other processes?) I am not sure what slurm does when you don't specify how much memory you need? Does it default to some particular amount or let you go over the limit, and then take it away if it needs it?

I have run SCAMP on SLURM before but never with more than one node. I'm not sure if that could be relevant. All of this should be invisible to SCAMP though.

Unfortunately, without having your exact setup and GPUs available to me I can't attempt to replicate. From now on, can you run with --print_debug_info it will cause a lot of output, but it might help us figure out what is going on.

It would also be useful to know which configurations it is failing in and if it is always failing in those configurations or only sometimes. If it isn't happening every time that points to the environment SCAMP is running in. SCAMP's behavior is mostly deterministic.

Another tip is if you have access to a Tesla GPU you are going to get significantly more performance due to the double precision Floating Point units. That is why your run is taking 2 days to finish. The K80s/Geforce Titan/P100/V100s are significantly faster because they have more double precision units.

**Comment by savardge** [comment]
Hey, so it looks like in the end the problem is more likely related to SLURM like we suspected. When requesting multiple nodes, the job was not actually running the code on the GPUs of all these nodes: I realized that after ssh'ing into each node, the resources were not being used... So it must be something about slurm not being able to distribute a GPU job to multiple nodes and getting confused if one tries to? 
Anyway, I've tried again using the same data with only one node and the code doesn't fail in that case, so I will just stick to that. :-)
Thanks!
